### ISSUE TITLE
Smarter allocation of ECS resources

### ISSUE THREAD

**Issue by abby-fuller** [aws/amazon-ecs-agent#225]
@euank 

Doesn't seem like ECS rebalances tasks to allocate resources more effectively.  For example,  if I have task A and task B running on different cluster hosts, and try to deploy task C, I'll get a resource error, and task C will fail to deploy.  However, if ECS rebalanced tasks A and B to run on the same box, there would be enough resources to deploy task C.

This comes up pretty often for us-  with the current placement of tasks, we run of resources and cannot deploy/achieve 100% utilization of our cluster hosts, because things are balanced inefficiently.  Max, we're probably getting 75% utilization, and 25% is going to waste.

**Comment by simplycloud** [comment]
Thanks for the report Abby. Yes, this is correct our scheduler does not currently re-balance tasks across an existing cluster. This is something that is on our roadmap.

**Comment by nathanpeck** [comment]
+1

This is a great example of the issue at its worst:

![screen shot 2015-10-15 at 1 52 03 pm](https://cloud.githubusercontent.com/assets/3495674/10522615/146a3d50-7344-11e5-8eee-5388441d1bf4.png)

Due to extremely poor balancing of the tasks by the scheduler I'm blocked from deploying a new revision of the `hyperspace` container. The only instance that has spare capacity for a new `hyperspace` container is already running a `hyperspace` container so the ECS agent just gives up.

The ideal thing to do would be for the ECS agent to automatically rebalance tasks from the first instance onto the second instance which has tons of extra resources, so that a new hyperspace could be placed onto the first instance.

Right now the only way to get ECS to operate smoothly is to allocate tons of extra server capacity so that there is always a huge margin of extra capacity for deploying new task definitions. I'd expect to need maybe say 10%-15% extra spare capacity, but it seems that things only work properly if there is nearly 50% spare capacity available.

**Comment by skarungan** [comment]
During a restart of services too, the same occurs.
![ecs schedule](https://cloud.githubusercontent.com/assets/13727764/15995016/aac5f3ca-3130-11e6-8878-82047cde44e4.png)

This needs me to run a spare instance, just for proper re-scheduling during restarts.
**Bigger Problem** 
This issue occurs on a large scale, which is very problematic. If I have 10 tasks with 1 vCPU and 1 task with 2vCPU and 10 t2.mediums (2 vCPU), the 10 smaller tasks get launched on each of the instance and 2vCPU gets not resource left (rare occurrence though), but hitting this 2 to 3 times a month

![ecs schedule big](https://cloud.githubusercontent.com/assets/13727764/15995069/f793aa8e-3131-11e6-88c0-e45a5f8fdde2.png)

Please prioritize this issue. 
Thanks.

**Comment by djenriquez** [comment]
#338 is requesting the same feature

**Comment by nathanpeck** [comment]
@skarungan @djenriquez ECS has added a new feature that kind of fixed this for us. You can specify a "minimum percentage" that will allow the agent to shut down old task definitions before starting new task definitions. So let's say you are running 6 of the old task definition and you want to deploy 6 of the new task definition but there isn't enough resources to add even 1 more copy of the new task defintion. If you have min percentage at 50% then ECS agent will kill 3 of the old tasks and start 3 of the new tasks, then once those 3 new tasks are started it kills off the remaining 3 old tasks and starts 3 more new tasks.

**Comment by djenriquez** [comment]
Ah @nathanpeck, I completely misunderstood the feature request here. Though what you describe is very helpful, I was actually looking for a "rebalance" feature where as new instances come up in the cluster, ECS will reevaluate its distribution of tasks.

**Comment by ishworgurung** [comment]
@simplycloud where are we at with this FR? ta.

**Comment by mwarkentin** [comment]
+1 for something smarter from ECS on this.

We're currently using Convox to manage our ECS clusters, and our racks are currently running on `t2` instances - we've noticed that tasks seem to cluster on a couple of instances, while others are pretty much empty. With the t2's in particular, this means that some instances are chewing through their CPU credits while others are gaining. 

Smarter balancing of tasks would help make this more feasible until we actually outgrow the t2 class.

**Comment by jfstephe** [comment]
@simplycloud can you please respond? This is biting people (myself included). It's been over a year since you mentioned this was on the road map. Where was it on the 15 Oct 2015 and where is it now?

**Comment by matthewford** [comment]
@mwarkentin I would not use convox with any of the t2 classes, there is a build agent that builds the images that will eat all your CPU credits on deployments.

**Comment by matthewford** [comment]
@mwarkentin ignore me I thought I was commenting on a convox issue.

**Comment by djenriquez** [comment]
Hi all, just hoping for an update on the rebalance feature. 

Ideally, this feature should handle something similar to [Nomad's System jobs](https://www.nomadproject.io/docs/runtime/schedulers.html), where it is just an event that ECS is listening to and can perform actions against.

**Comment by thomask** [comment]
+1

**Comment by Zetanova** [comment]
+1 rebalancing wih the include of credit balance would be good.
I would already be happy if an EC2 container instance would be drained when 
it has low credit balance.

**Comment by kevindiamond** [comment]
@samuelkarp @aaithal or others:
Any plans to add this feature or something similar?

**Comment by debu99** [comment]
same problem as mine...

<img width="1096" alt="screen shot 2017-07-06 at 21 44 13" src="https://user-images.githubusercontent.com/25805758/27914159-e5933fb2-6294-11e7-8e45-ac2f23f9393c.png">
<img width="1008" alt="screen shot 2017-07-06 at 21 49 24" src="https://user-images.githubusercontent.com/25805758/27914204-0819aaf8-6295-11e7-89c0-5942f2e164e9.png">

**Comment by nathanpeck** [comment]
@debu99 Sorry its not clear to me, what is the issue in your screenshots?

**Comment by mwarkentin** [comment]
@nathanpeck I think that 6/7 tasks are running on the same instance (and the one that's not looks like a daemon that runs on all instances).

**Comment by devshorts** [comment]
+1 on rebalancing when the cluster grows

**Comment by oogali** [comment]
My two cents...

I'd imagine that it'd be unwanted behavior to include by default, so it should be something that the user should opt-in to on a per-service basis -- if you're running a web app behind a load balancer or anything ephemeral, it's relatively safe to "move" these.

(By "move", I mean start a new instance on the new container instance, without the count of tasks in the RUNNING/PENDING state exceeding the maximum percent; and once it's running, kill the old task)

However, the rebalancing should *not* take the limited CPU credit of throttled instance types (tX) into account. To effectively do that, you'd need to generate a large volume of CloudWatch statistics gathering requests and/or estimate the rate of depletion which will push up your overall bill. By how much? I'm not sure, but it will start to chip away at the cost advantage you're intended to have by using the tX instance types.

You're better off using non-throttled instances for ECS clusters, or you'll need to roll your own logic/framework for managing CPU credit balances. (One can argue this is the standard "buy-versus-build" argument)

As far as I'm aware, throttled instance types are meant for burstable workloads that can be terminated and restarted without incurring pain or major service degradation. Your containers may be considered a burstable workload, but the ECS agent and Docker daemon are not -- if you starve these two of CPU, your entire ECS instance may become unusable/invisible to the ECS scheduler.

Lastly, this would not necessarily something that would live within the ECS agent, but instead need to be a centrally coordinated service within the larger EC2 Container Service.

Otherwise, the sole complexity cost of having ECS agents be aware of other ECS agents (agent-to-agent discovery, heartbeats to make sure the other ECS agent hasn't died, communication and coordination patterns to request-accept-send tasks, limiting broadcast traffic to not propagate beyond a single cluster, and many other distributed computing concerns) would make this feature DOA before you even get to solving the problem of **how** to properly rebalance.

**Comment by Zetanova** [comment]
I would already be happy if an EC2 container instance would be drained when
it has a low credit balance and activated on a specific higher limit.

The autoscale feature could manage the rest to add more EC2 instances.

The main problem is just that if an tX instance is out of credits,
it is in ecs online and healthy node/container
but in real it is a "unhealthy" node that cant surf all services

**Comment by darrylb-github** [comment]
+1 for rebalancing

**Comment by amitsaxena** [comment]
Wouldn't using 'binpack' task placement strategy help with rebalancing in the scenarios described above. It's described in more detail here:
http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-placement-strategies.html

binpack makes sure that the instance is utilized to the maximum before placing the task on a fresh node. I understand that there will be edge cases where task placement will still not be as desired, but I think this will solve the majority of problems for you.

**Comment by devshorts** [comment]
@amitsaxena I'm not sure that solves the problem at all. Binpack isn't HA since it will try and pack as many items on the box as possible. 

The issue in this ticket is that you want to have even _spread_. So as you add more nodes to the cluster you want to maximize HA by moving tasks off of nodes that are more heavily loaded onto newer nodes that are less heavy loaded. This way each node is utilized, vs having some nodes that are running at near peak and other nodes that may have zero or very few services. This is the situation we run into now, we have nodes that are running at around 75% capacity and we'd like to make sure they have breathing room to bursts.  If we add a new node to the cluster we now have a node that is doing _nothing_ vs having nodes that are at our peak run less hot and have more headroom.

**Comment by thomasbiddle** [comment]
Anyone working on this? It's currently the second highest rated FR. I'm hesitant to even use ECS due to this.

**Comment by jfstephe** [comment]
I wonder if the new [EKS service](https://aws.amazon.com/eks/) makes this less of an issue, or at least provides an option (or explanation) for the lack of progress on this issue? Some guidance from AWS would be great! 
@simplycloud @samuelkarp @aaithal @anyone with some information! Please respond.

**Comment by mwarkentin** [comment]
Watching the reinvent keynote and @abby-fuller just came on.. then opened this issue because of a notification.. hah!

**Comment by briziomusic** [comment]
+1 for rebalancing

**Comment by amitsaxena** [comment]
@devshorts I sort of get your point about having an even spread, but having binpack placement strategy along with Multi AZ is possible as well. Something similar to:

```
"placementStrategy": [
    {
        "field": "attribute:ecs.availability-zone",
        "type": "spread”
    },
    {
        "field": "memory",
        "type": "binpack"
    }
]
```

So I don't think HA is a problem with binpack strategy, and it can be effectively used to have maximum utilization of instances before spawning new ones.

PS: Sorry for the late reply. I somehow missed your original comment :)

**Comment by pparth** [comment]
+1 on rebalancing when the cluster grows.

**Comment by soukicz** [comment]
If there was a way to drain&stop task, we could implement this in Lambda but there is no way to do that without hacking every health check.

**Comment by azaytsev-csr** [comment]
For me the biggest issue has been when the VM is removed from a cluster (say for a 2 VM cluster) and then re-added (e.g. simulating an AZ outage), i no longer get real HA if i spread just across 2 AZs, because all the tasks are now on one instance in one AZ.

**Comment by lodotek** [comment]
+1

**Comment by relu** [comment]
Any updates on this?

**Comment by mrapczynski** [comment]
@PaulLiang1 Good find, and a neat trick with Lambda.

However, I agree with the overall sentiment of this thread that at least a basic version of rebalancing ought to be built-in. Such functionality goes hand-in-hand with other AWS services like auto scaling groups where I have container instances coming and going entirely on their own through policies.

Writing Lambdas on an event stream should be for when we need a special scaling behavior for unique application or situation.

Auto rebalancing when registered instances change is something **everybody** can use, and shouldn't be a bespoke addition to the configuration.

**Comment by Puneeth-n** [comment]
> this might help
> https://github.com/aws-samples/ecs-refarch-task-rebalancing#rebalance

@mrapczynski @PaulLiang1 The above lambda can be made way simpler by just forcing a new deployment of the ECS service without creating a new task definition.

But I agree, It would be cool if AWS did it automagically! Hoping they do it just like the introduced `DAEMON` ECS services :)

**Comment by ricardosllm** [comment]
+1

**Comment by psyhomb** [comment]
+1

**Comment by momcilovic** [comment]
+1

**Comment by RealCoderiver** [comment]
+1

**Comment by psyhomb** [comment]
> Thanks for the report Abby. Yes, this is correct our scheduler does not currently re-balance tasks across an existing cluster. This is something that is on our roadmap.

@simplycloud this is something that was on the ECS roadmap 3 years ago, what's the status 3 years later?

I would also add that it would be absolutely more fair admitting that we all shoud start using EKS and that this or similar features are never going to be implemented nor available in ECS.

**Comment by Puneeth-n** [comment]
AZ spread && binpack  (In that order) seems to do the trick for me. There are no wastage of resources (unless you provisioned a bit of headroom for scaling)

![screen shot 2018-08-09 at 11 38 15 am](https://user-images.githubusercontent.com/5619473/43891126-e187c3de-9bc8-11e8-9300-1adc7878847b.png)

**Comment by aric49** [comment]
+1 

This would be super helpful. We are having issues with ECS scheduling tasks on pegged instances while ignoring newly created instances.

**Comment by francoisvdv** [comment]
Same issue here, where certain instances are bogged down with tasks while others are empty. Could you provide a status update on this feature?

**Comment by dsouzajude** [comment]
+1

**Comment by leonfs** [comment]
+1

**Comment by dsouzajude** [comment]
For some of the use-cases mentioned above, in particular rebalancing tasks across container instances in a cluster, and also rebalancing tasks when the cluster (i.e. the autoscaling group) scales up, this has very much helped me a lot and is a much simpler solution.

https://github.com/maginetv/ecs-task-balancer

Would love your feedback on this.

**Comment by hehnope** [comment]
With EKS and all, is this still a serious feature? Rebalancing is basically a requirement when resizing/downsizing a cluster.

**Comment by coultn** [comment]
Thanks everyone for the feedback on this issue.  I wanted to let you know that we on the ECS team are aware of this issue, and that it is under active consideration.  +1's and additional details on use cases are always appreciated and will help inform our work moving forward.

**Comment by psyhomb** [comment]
Is there anything new on this issue?

**Comment by coultn** [comment]
@psyhomb We are looking at potential solutions.  I cannot share more details at this time, but please be assured that we are actively working on it.

**Comment by plektra** [comment]
+1 for the sake of re:Invent!

**Comment by matthewjohnson2821** [comment]
+1 on rebalancing

**Comment by ajain04** [comment]
I am trying to setup what is mentioned in this blog post: https://aws.amazon.com/blogs/compute/how-to-automate-container-instance-draining-in-amazon-ecs/

I am facing issue with the re-balancing of the containers. I have an ECS cluster with 2 instances and service with 2 containers. The task placement strategy I am using AZ spread. When we deploy or update the service , the containers are placed across AZs but when we are updating the cluster image and change instances, the containers are placed on 1 instance within the same AZ.

![image](https://user-images.githubusercontent.com/29250679/49272740-a4afef80-f498-11e8-9bc2-7a1afb07bbd8.png)

![image](https://user-images.githubusercontent.com/29250679/49272773-babdb000-f498-11e8-90e0-936e2a360c1e.png)

**Comment by psyhomb** [comment]
@coultn 
>We are looking at potential solutions. I cannot share more details at this time, but please be assured that we are actively working on it.

If you're actively working on this, like you said, I would like to know why is this issue still on the Proposed list then?
https://github.com/aws/containers-roadmap/labels/Proposed

**Comment by coultn** [comment]
@psyhomb we are looking closely at this problem, and trying to determine a solution that makes the most sense.  As you can see even from this comment thread, there are many different aspects to the problem and potential solutions.  I'm interested to hear what is most important to you - what specific problem are you trying to solve, and what would you need to see in a solution?

**Comment by psyhomb** [comment]
@coultn given that more or less Amazon is taking almost everything from the OSS please revisit some of the OSS projects, like for example ClusterLabs'  [pacemaker](http://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/2.0/html/Pacemaker_Explained/ch11.html), and maybe you will find potential solution for the above explained issue, instead of asking generic questions!

**Comment by hehnope** [comment]
@coultn since you're asking, the problem we are running into is two fold:

1. If we increase the size of the cluster, then nodes for hours can be sitting with no tasks.
2. If we decrease the size of the cluster, then the scheduler seems to favor specific nodes for dumping tasks on.

So, with regards to solutions:

1. To address the first problem; I'd imagine a better solution would be to check which nodes in the cluster at an interval are "unbalanced"; and then move tasks to those nodes to balance the claster.
2. To address the second problem; instead of "dumping" tasks on a few target nodes when a cluster shrinks, then some kind of check should be done before task placement to make sure one node isnt just getting overloaded with new tasks.

**Comment by jamiegs** [comment]
If you have multiple ECS services that are autoscaling through the day you might want to re-pack the containers and and scale-in the hosts to remove some excess capacity. 

But it should to be smart about which container host it should terminate for a scale-in.. you'd want to keep zones balanced and terminate the host with the least amount of resources reserved.. otherwise it's going to have to relocate many containers.

Currently the ASG isn't smart about which host it terminates on scale-in and usually kills the oldest host, which will often be the one with the most running containers.

**Comment by abby-fuller** [comment]
moving this over to the [container roadmap](https://github.com/aws/containers-roadmap/projects/1), since it's ultimately a set of feature requests, and not an ecs-agent bug.
